File: game/Player.py


from typing import TypeVar


from database.queries.types import get_resource_types
from Enum import Enum


Player = TypeVar("Player")
Players = list[Player]
Port = TypeVar("Port")
Ports = list[Port]
Road = TypeVar("Road")
Roads = list[Road]
Settlement = TypeVar("Settlement")
Settlements = list[Settlement]


class Player:
	ResourcesTypes = Enum("Player::ResourceType", **{type["label"]: type["id"]-1 for type in get_resource_types()})

	def __init__(self, id: int):
		self.id: int = id

		self.resources = []  # TODO
		self.roads: list[Road] = []
		self.settlements: list[Settlement] = []


	def __eq__(self, right: Player) -> bool:
		return self.id == right.id


	def ports(self) -> list[Port]:
		found_ports: list[Port] = []
		for settlement in self.settlements:
			for port in settlement.ports:
				if(port not in found_ports):
					found_ports.append(port)

		return found_ports

File: Source/webpage/Scripts.ts/Game/Player.ts


import { Road, Settlement } from "./Board/index.js";


import { ResourceCount, ResourceType } from "../Types.d.js";


export class Player
{
	id: number;
	name: string;
	development_cards: ResourceCount[];  // TODO: add correct type
	resources: ResourceCount[];

	roads: [Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null, Road|null];
	settlements: [Settlement|null, Settlement|null, Settlement|null, Settlement|null, Settlement|null, Settlement|null, Settlement|null, Settlement|null, Settlement|null];


	constructor(id: number, name: string, development_cards: ResourceCount[], resources: ResourceCount[])
	{
		this.id = id;
		this.name = name;
		this.development_cards = development_cards;
		this.resources = resources;
		this.roads = [null, null, null, null, null, null, null, null, null, null, null, null, null, null, null];
		this.settlements = [null, null, null, null, null, null, null, null, null];
	}
}
